sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -701,3 +702,43 @@ object RewriteLateralSubquery extends Rule[LogicalPlan] {
       Join(left, newRight, joinType, newCond, JoinHint.NONE)
   }
 }
+
+/**
+ * This rule optimizes subqueries with OneRowRelation as leaf nodes.
+ */
+object OptimizeOneRowRelationSubquery extends Rule[LogicalPlan] {
+
+  object OneRowSubquery {
+    def unapply(plan: LogicalPlan): Option[Seq[NamedExpression]] = {
+      CollapseProject(EliminateSubqueryAliases(plan)) match {
+        case Project(projectList, _: OneRowRelation) => Some(stripOuterReferences(projectList))
+        case _ => None
+      }
+    }
+  }
+
+  /**
+   * Rewrite a subquery expression into one or more expressions. The rewrite can only be done
+   * if there is no nested subqueries in the subquery plan.
+   */
+  private def rewrite(plan: LogicalPlan): LogicalPlan = plan.transformUpWithSubqueries {
```

Comment:
do we need to handle nested subqueries here? I think the rule `OptimizeSubqueries` will run this rule again to optimize nested subqueries.

Reply:
The reason why we need to check subqueries is to deal with nested subqueries:
```
Project [scalar-subquery [a]]
:  +- Project [scalar-subquery [b]] <-- collapsible if transform with nested subqueries first
:     :  +- Project [outer(b) + 1]
:     :     +- OneRowRelation
:     +- Project [outer(a) as b]
:         +- OneRowRelation
+- Relation [a]
```
A subquery's plan should only be rewritten if it doesn't contain another correlated subquery. If we do not transform the nested subqueries first, we will miss out cases like the one above.